### PR TITLE
WIP: EscapeAnalysis: Implement CodeInstance caching

### PR DESCRIPTION
**Pull request by ianatol**
This seems to be the way we discussed caching, but I noticed a couple problems:

1. Building is really slow:
```
Sysimage built. Summary:
Total ───────  73.943191 seconds 
Base: ───────  30.145973 seconds 40.7691%
Stdlibs: ────  43.795525 seconds 59.2286%
    JULIA usr/lib/julia/sys-o.a
Generating REPL precompile statements... 36/36
Executing precompile statements... 1310/1339
Precompilation complete. Summary:
Total ─────── 167.419110 seconds
Generation ── 118.754552 seconds 70.9325%
Execution ───  48.664558 seconds 29.0675%
```
2. Breaks some EscapeAnalysis tests (see below)

Certainly possible I missed something when changing `CodeInstance`, but we may need to do a bit more work here
CC: @aviatesk

### CODE REVIEW

File: base/compiler/typeinfer.jl [hunk 1]
```diff
@@ -389,7 +389,7 @@ function cache_result!(interp::AbstractInterpreter, result::InferenceResult)
     end
 
     if isdefined(result, :escapes)
-        EscapeAnalysis.GLOBAL_ESCAPE_CACHE[linfo]= result.escapes
+        code_cache(interp)[linfo].arg_escapes = result.escapes
```
Comment (vchuravy) [inline base/compiler/typeinfer.jl:392]: Where is this read? I only see `arg_escapes` on the LHS

Comment (vchuravy) [reply base/compiler/typeinfer.jl:392]: Yeah it seems to be you need to change `EscapeAnalysis.jl:1222` to actually read this and then delete line 592

Comment (aviatesk) [reply base/compiler/typeinfer.jl:392]: yes.

And it would be clearer if we just pass `escapes` to the `CodeInstance` constructor.

**Comment by aviatesk** [comment]
I think we also need to implement a support for serialization.

**Comment by aviatesk** [comment]
close in favor of 91204ed094e208136045324f32d171074cfa2132

**Pull request closed without merging**